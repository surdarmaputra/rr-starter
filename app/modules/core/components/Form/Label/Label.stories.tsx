import type { Meta, StoryObj } from '@storybook/react';

import { Label, type LabelProps } from './Label';

type Story = StoryObj<typeof Label>;

const meta: Meta = {
  component: Label,
  title: 'Core/Components/Form/Label',
};

export default meta;

export const Default: Story = {
  render: (args: LabelProps) => <Label {...args}>Sample label</Label>,
};
