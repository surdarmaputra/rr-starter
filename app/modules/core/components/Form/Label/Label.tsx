import * as LabelPrimitive from '@radix-ui/react-label';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '~/modules/core/libs/shadcn/utils';

const labelVariants = cva(
  'text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-letter-body',
);

export interface LabelProps
  extends React.ComponentPropsWithoutRef<typeof LabelPrimitive.Root>,
    VariantProps<typeof labelVariants> {
  secondaryText?: React.ReactNode;
}

const Label = React.forwardRef<
  React.ComponentRef<typeof LabelPrimitive.Root>,
  LabelProps
>(({ children, className, secondaryText, ...props }, ref) => (
  <LabelPrimitive.Root
    className={cn(labelVariants(), className)}
    ref={ref}
    {...props}
  >
    {children}
    {Boolean(secondaryText) && (
      <span className="ml-1.5 font-normal text-letter-caption">
        {secondaryText}
      </span>
    )}
  </LabelPrimitive.Root>
));
Label.displayName = LabelPrimitive.Root.displayName;

export { Label };
