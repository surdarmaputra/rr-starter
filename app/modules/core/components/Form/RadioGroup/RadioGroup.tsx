import * as RadioGroupPrimitive from '@radix-ui/react-radio-group';
import { Circle } from 'lucide-react';
import * as React from 'react';

import { cn } from '~/modules/core/libs/shadcn/utils';

export type RadioGroupProps = React.ComponentPropsWithoutRef<
  typeof RadioGroupPrimitive.Root
>;

const RadioGroup = React.forwardRef<
  React.ComponentRef<typeof RadioGroupPrimitive.Root>,
  RadioGroupProps
>(({ className, ...props }, ref) => {
  return (
    <RadioGroupPrimitive.Root
      className={cn('grid gap-2', className)}
      {...props}
      ref={ref}
    />
  );
});
RadioGroup.displayName = RadioGroupPrimitive.Root.displayName;

const RadioGroupItem = React.forwardRef<
  React.ComponentRef<typeof RadioGroupPrimitive.Item>,
  React.ComponentPropsWithoutRef<typeof RadioGroupPrimitive.Item>
>(({ className, ...props }, ref) => {
  return (
    <RadioGroupPrimitive.Item
      className={cn(
        'aspect-square h-4 w-4 rounded-full border border-secondary ring-offset-white',
        'focus:outline-none focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2',
        'disabled:cursor-not-allowed disabled:opacity-50',
        'data-[state=checked]:border-primary',
        className,
      )}
      ref={ref}
      {...props}
    >
      <RadioGroupPrimitive.Indicator className="flex items-center justify-center">
        <Circle className="h-2.5 w-2.5 fill-primary text-primary" />
      </RadioGroupPrimitive.Indicator>
    </RadioGroupPrimitive.Item>
  );
});
RadioGroupItem.displayName = RadioGroupPrimitive.Item.displayName;

export { RadioGroup, RadioGroupItem };
