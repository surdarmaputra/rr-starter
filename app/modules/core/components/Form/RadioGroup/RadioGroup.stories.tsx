import type { Meta, StoryObj } from '@storybook/react';

import { Label } from '../Label/Label';
import { RadioGroup, RadioGroupItem, type RadioGroupProps } from './RadioGroup';

type Story = StoryObj<typeof RadioGroup>;

const meta: Meta = {
  component: RadioGroup,
  title: 'Core/Components/Form/RadioGroup',
};

export default meta;

export const Default: Story = {
  args: {
    defaultValue: 'option2',
  },
  render: (args: RadioGroupProps) => (
    <RadioGroup {...args}>
      <div className="flex items-center space-x-2">
        <RadioGroupItem id="option1" value="option1" />
        <Label htmlFor="option1">Option 1</Label>
      </div>
      <div className="flex items-center space-x-2">
        <RadioGroupItem id="option2" value="option2" />
        <Label htmlFor="option2">Option 2</Label>
      </div>
    </RadioGroup>
  ),
};
