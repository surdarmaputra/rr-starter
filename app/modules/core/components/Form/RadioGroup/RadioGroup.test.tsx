import {
  fireEvent,
  render,
  screen,
  waitFor,
  within,
} from '@testing-library/react';
import { beforeEach, describe, expect, test, vi } from 'vitest';

import { Label } from '../Label/Label';
import { RadioGroup, RadioGroupItem, type RadioGroupProps } from './RadioGroup';

describe('RadioGroup', () => {
  function renderComponent(props?: RadioGroupProps) {
    return render(
      <RadioGroup
        data-testid="RadioGroup-testid"
        defaultValue="option2"
        {...props}
      >
        <div className="flex items-center space-x-2">
          <RadioGroupItem id="option1" value="option1" />
          <Label htmlFor="option1">Option 1</Label>
        </div>
        <div className="flex items-center space-x-2">
          <RadioGroupItem id="option2" value="option2" />
          <Label htmlFor="option2">Option 2</Label>
        </div>
      </RadioGroup>,
    );
  }

  beforeEach(() => {
    vi.clearAllMocks();
  });

  test('renders default value', async () => {
    renderComponent();
    const radioGroup = await screen.findByTestId('RadioGroup-testid');
    const options = within(radioGroup).getAllByRole('radio');
    expect(options).toHaveLength(2);
    await waitFor(() => expect(options[1]).toBeChecked());
    await waitFor(() => expect(options[0]).not.toBeChecked());
  });

  test('triggers onValueChange', async () => {
    const handleChange = vi.fn();
    renderComponent({ onValueChange: handleChange });
    await waitFor(() =>
      expect(screen.getByRole('radio', { name: /Option 2/ })).toBeChecked(),
    );
    fireEvent.click(screen.getByRole('radio', { name: /Option 1/ }));
    await waitFor(() => expect(handleChange).toHaveBeenCalledTimes(1));
    expect(handleChange).toHaveBeenCalledWith('option1');
    await waitFor(() =>
      expect(screen.getByRole('radio', { name: /Option 1/ })).toBeChecked(),
    );
  });

  test('renders snapshot', async () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
