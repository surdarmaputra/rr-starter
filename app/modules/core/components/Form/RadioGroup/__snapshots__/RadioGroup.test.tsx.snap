// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RadioGroup > renders snapshot 1`] = `
<div>
  <div
    aria-required="false"
    class="grid gap-2"
    data-testid="RadioGroup-testid"
    dir="ltr"
    role="radiogroup"
    style="outline-color: none; outline-style: none; outline-width: initial;"
    tabindex="0"
  >
    <div
      class="flex items-center space-x-2"
    >
      <button
        aria-checked="false"
        class="aspect-square h-4 w-4 rounded-full border border-secondary ring-offset-white focus:outline-none focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:border-primary"
        data-radix-collection-item=""
        data-state="unchecked"
        id="option1"
        role="radio"
        tabindex="-1"
        type="button"
        value="option1"
      />
      <label
        class="text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-letter-body"
        for="option1"
      >
        Option 1
      </label>
    </div>
    <div
      class="flex items-center space-x-2"
    >
      <button
        aria-checked="true"
        class="aspect-square h-4 w-4 rounded-full border border-secondary ring-offset-white focus:outline-none focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:border-primary"
        data-radix-collection-item=""
        data-state="checked"
        id="option2"
        role="radio"
        tabindex="-1"
        type="button"
        value="option2"
      >
        <span
          class="flex items-center justify-center"
          data-state="checked"
        >
          <svg
            class="lucide lucide-circle h-2.5 w-2.5 fill-primary text-primary"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
          </svg>
        </span>
      </button>
      <label
        class="text-sm leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 text-letter-body"
        for="option2"
      >
        Option 2
      </label>
    </div>
  </div>
</div>
`;
